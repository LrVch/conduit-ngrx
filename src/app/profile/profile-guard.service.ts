import { Injectable } from '@angular/core';
import {
    CanActivate,
    ActivatedRouteSnapshot
} from '@angular/router';
import { Observable, of, throwError } from 'rxjs';
import { Store } from '@ngrx/store';
import { AppState } from '../reducers';
import { Router } from '@angular/router';
import { tap, take, map, switchMap, catchError } from 'rxjs/operators';
import { ProfilesService } from '../core';
import { ProfileLoadSuccess } from './profile.actions';

@Injectable()
export class ProfileGuard implements CanActivate {
    NO_USER_NAME = 'No user name';
    constructor(
        private store: Store<AppState>,
        private router: Router,
        private profileService: ProfilesService
    ) {

    }
    canActivate(
        route: ActivatedRouteSnapshot
    ): Observable<boolean> {

        const username = route.params['username'];

        return this.store.pipe(
            take(1),
            switchMap(() => username ? this.profileService.get(username) : throwError(this.NO_USER_NAME)),
            map(profile => new ProfileLoadSuccess({ profile })),
            tap((action: ProfileLoadSuccess) => this.store.dispatch(action)),
            map(profile => !!profile),
            catchError((error) => {
                console.error(error);
                this.router.navigate(['/']);
                return of(false);
            })
        );
    }
}
