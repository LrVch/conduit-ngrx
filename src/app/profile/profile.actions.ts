import { Action } from '@ngrx/store';
import { Profile } from '@app/core';

export enum ProfileActionTypes {
  ProfileLoadRequest = '[Profile] Profile Load Request',
  ProfileLoadSuccess = '[Profile] Profile Load Success',
  ProfileLoadFail = '[Profile] Profile Load Fail',
  ProfileClear = '[Profile] Profile Clear',

  ProfileToggleFollowingRequest = '[Profile] Profile Toggle Following Request',
  ProfileToggleFollowingSuccess = '[Profile] Profile Toggle Following Success',
  ProfileToggleFollowingFail = '[Profile] Profile Toggle Following Fail',
  ProfileFollowingRequest = '[Profile] Profile Following Request',
  ProfileUnFollowingRequest = '[Profile] Profile UnFollowing Request',

  SetFollowingProfile = '[Profile] Set Following Profile',
  ClearFollowingProfile = '[Profile] Clear Following Profile'
}

export class ProfileLoadRequest implements Action {
  readonly type = ProfileActionTypes.ProfileLoadRequest;

  constructor(public payload: { username: string }) { }
}

export class ProfileLoadSuccess implements Action {
  readonly type = ProfileActionTypes.ProfileLoadSuccess;

  constructor(public payload: { profile: Profile }) { }
}

export class ProfileLoadFail implements Action {
  readonly type = ProfileActionTypes.ProfileLoadFail;
}

export class ProfileClear implements Action {
  readonly type = ProfileActionTypes.ProfileClear;
}

export class ProfileToggleFollowingRequest implements Action {
  readonly type = ProfileActionTypes.ProfileToggleFollowingRequest;

  constructor(public payload: { profile: Profile }) { }
}

export class ProfileToggleFollowingSuccess implements Action {
  readonly type = ProfileActionTypes.ProfileToggleFollowingSuccess;

  constructor(public payload: { profile: Profile, showNotification?: boolean }) { }
}

export class ProfileToggleFollowingFail implements Action {
  readonly type = ProfileActionTypes.ProfileToggleFollowingFail;

  constructor(public payload?: { profile: Profile, showNotification?: boolean }) { }
}

export class ProfileFollowingRequest implements Action {
  readonly type = ProfileActionTypes.ProfileFollowingRequest;

  constructor(public payload: { profile: Profile }) { }
}

export class ProfileUnFollowingRequest implements Action {
  readonly type = ProfileActionTypes.ProfileUnFollowingRequest;

  constructor(public payload: { profile: Profile }) { }
}

export class SetFollowingProfile implements Action {
  readonly type = ProfileActionTypes.SetFollowingProfile;

  constructor(public payload: { profile: Profile }) { }
}

export class ClearFollowingProfile implements Action {
  readonly type = ProfileActionTypes.ClearFollowingProfile;
}

export type ProfileActions =
  ProfileLoadRequest
  | ProfileLoadSuccess
  | ProfileLoadFail
  | ProfileClear
  | ProfileToggleFollowingRequest
  | ProfileToggleFollowingSuccess
  | ProfileToggleFollowingFail
  | ProfileFollowingRequest
  | ProfileUnFollowingRequest
  | SetFollowingProfile
  | ClearFollowingProfile;
