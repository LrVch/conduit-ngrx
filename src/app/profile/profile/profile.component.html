<div class="profile-page">

  <app-user-info (folowing)="onFolowing($event)"
                 [isUser]="isUser$ | async"
                 [profile]="profile$ | async"></app-user-info>

  <div class="content">

    <div class="container">
      <div class="row">
        <div class="col-xs-12 col-md-10 offset-md-1">
          <app-content-tabs [type]="type$ | async"
                            [tabs]="tabs$ | async"
                            (selectedType)="onSelectedType($event)">
          </app-content-tabs>
        </div>
      </div>
    </div>

    <div class="container">
      <div class="row">
        <div class="col-xs-12 col-md-10 offset-md-1">
          <div class="articles">
            <app-articles-list [isLoading]="loadingArticles$ | async"
                               [locale]="locale$ | async"
                               [contentLength]="limit$ | async"
                               [isErrorLoading]="errorLoadingArticles$ | async"
                               (favorited)="onFavorited($event)"
                               [articlesList]="atricles$ | async"></app-articles-list>
          </div>
        </div>
      </div>

    </div>
    <div class="container">
      <div class="row">
        <div class="col-xs-12 col-md-10 offset-md-1">
          <mat-paginator [hidden]="(articlesCount$ | async) < 1 || (loadingArticles$ | async)"
                         [length]="articlesCount$ | async"
                         (page)="onPageChange($event)"
                         [pageSize]="limit$ | async"
                         [pageIndex]="pageIndex$ | async"
                         [showFirstLastButtons]="true"
                         [pageSizeOptions]="pageSizeOptions">
          </mat-paginator>
        </div>
      </div>
      <div class="pagination">
      </div>
    </div>
  </div>