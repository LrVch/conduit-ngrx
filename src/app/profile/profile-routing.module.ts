import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { AuthGuard } from '@app/core';
import { ProfileComponent } from './profile/profile.component';
import { ProfileGuard } from './profile-guard.service';

const routes: Routes = [
  {
    path: '',
    component: ProfileComponent,
    canActivate: [ProfileGuard],
    data: { title: 'conduit.menu.profile' }
  },
  {
    path: ':username',
    component: ProfileComponent,
    canActivate: [ProfileGuard],
    data: { title: 'conduit.menu.profile' }
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class ProfileRoutingModule {}
