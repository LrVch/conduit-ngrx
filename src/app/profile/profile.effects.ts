import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { Router } from '@angular/router';
import {
  ProfileActionTypes,
  ProfileToggleFollowingRequest,
  ProfileUnFollowingRequest,
  ProfileFollowingRequest,
  ProfileToggleFollowingSuccess,
  ProfileToggleFollowingFail,
  SetFollowingProfile,
  ClearFollowingProfile
} from './profile.actions';
import { retry, catchError, map, filter, mergeMap, withLatestFrom, tap } from 'rxjs/operators';
import { ProfilesService, Profile } from '../core';
import { of, Observable } from 'rxjs';
import { selectAuthLoggedIn } from '../auth/auth.selectors';
import { LogoutAction, ClearReturnStateFromRouteChange, AuthActionTypes, LoginSuccess } from '../auth/auth.actions';
import { selectFollowingProfile } from './profile.selectors';
import { NotificationService } from '../core/services/notification.service';


@Injectable()
export class ProfileEffects {

  constructor(
    private actions$: Actions,
    private store: Store<AppState>,
    private profileService: ProfilesService,
    private notificationService: NotificationService
  ) { }

  @Effect()
  taggleFollowUserProfile$ = this.actions$.pipe(
    ofType<ProfileToggleFollowingRequest>(ProfileActionTypes.ProfileToggleFollowingRequest),
    withLatestFrom(this.store.select(selectAuthLoggedIn)),
    mergeMap(([action, isLoggedIn]): Observable<ProfileUnFollowingRequest | ProfileFollowingRequest | SetFollowingProfile> => {
      if (!isLoggedIn) {
        return of(new SetFollowingProfile({ profile: action.payload.profile }));
      }
      const { profile } = action.payload;
      const { following } = profile;

      if (following) {
        return of(new ProfileUnFollowingRequest({ profile }));
      } else {
        return of(new ProfileFollowingRequest({ profile }));
      }
    })
  );

  @Effect()
  setFollowingProfile$ = this.actions$.pipe(
    ofType<SetFollowingProfile>(ProfileActionTypes.SetFollowingProfile),
    map(() => new LogoutAction())
  );

  @Effect()
  clearReturnStateFromRouteChange$ = this.actions$.pipe(
    ofType<ClearReturnStateFromRouteChange>(AuthActionTypes.ClearReturnStateFromRouteChange),
    map(() => new ClearFollowingProfile())
  );

  @Effect()
  loginSuccessProfile$ = this.actions$.pipe(
    ofType<LoginSuccess>(AuthActionTypes.LoginSuccess),
    withLatestFrom(this.store.select(selectFollowingProfile)),
    filter(([action, profile]) => !!profile),
    map(([action, profile]) => new ProfileFollowingRequest({ profile }))
  );

  @Effect()
  profileFollowing$ = this.actions$.pipe(
    ofType<ProfileFollowingRequest>(ProfileActionTypes.ProfileFollowingRequest),
    withLatestFrom(this.store.select(selectFollowingProfile)),
    mergeMap(([action, favoritingProfile]) => {
      const { profile } = action.payload;
      const { username } = profile;

      return this.profileService.follow(username)
        .pipe(
          map(result => {
            return new ProfileToggleFollowingSuccess({ profile: result.profile, showNotification: !favoritingProfile });
          }),
          retry(3),
          catchError(error => {
            console.error(error);
            return of(new ProfileToggleFollowingFail({showNotification: !favoritingProfile}));
          })
        );
    })
  );

  @Effect()
  profileUnFollowing$ = this.actions$.pipe(
    ofType<ProfileUnFollowingRequest>(ProfileActionTypes.ProfileUnFollowingRequest),
    mergeMap((action) => {
      const { profile } = action.payload;
      const { username } = profile;

      return this.profileService.unfollow(username)
        .pipe(
          map(result => {
            return new ProfileToggleFollowingSuccess({ profile: result.profile });
          }),
          retry(3),
          catchError(error => {
            console.error(error);
            return of(new ProfileToggleFollowingFail());
          })
        );
    })
  );

  @Effect({ dispatch: false })
  profileToggleFollowingSuccess$ = this.actions$.pipe(
    ofType<ProfileToggleFollowingSuccess>(ProfileActionTypes.ProfileToggleFollowingSuccess),
    filter(action => action.payload.showNotification),
    map(action => action.payload.profile),
    tap((profile: Profile) => this.notificationService.success({ message: 'Added to your favorites authors' }))
  );

  @Effect({ dispatch: false })
  profileToggleFollowingFail$ = this.actions$.pipe(
    ofType<ProfileToggleFollowingFail>(ProfileActionTypes.ProfileToggleFollowingFail),
    filter(action => action.payload.showNotification),
    tap(() => this.notificationService.error({ message: 'Can\'t add author to your favorites' }))
  );
}
