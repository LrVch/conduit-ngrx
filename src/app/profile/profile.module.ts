import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ProfileComponent } from './profile/profile.component';
import { ProfileRoutingModule } from './profile-routing.module';
import { SharedModule } from '@app/shared';
import { EffectsModule } from '@ngrx/effects';
import { ProfileEffects } from './profile.effects';
import { StoreModule } from '@ngrx/store';
import * as fromProfile from './profile.reducer';
import { ProfileGuard } from './profile-guard.service';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { HttpClient } from '@angular/common/http';
import { environment } from '@env/environment';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    ProfileRoutingModule,
    EffectsModule.forFeature([ProfileEffects]),
    StoreModule.forFeature('profile', fromProfile.profileReducer),
    TranslateModule.forChild({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient]
      }
    })
  ],
  declarations: [ProfileComponent],
  providers: [ProfileGuard]
})
export class ProfileModule {}

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.i18nPrefix}/assets/i18n/`,
    '.json'
  );
}
