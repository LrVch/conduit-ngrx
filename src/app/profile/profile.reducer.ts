import { Profile } from '@app/core';
import { ProfileActions, ProfileActionTypes } from './profile.actions';

export interface ProfileState {
  profile: Profile;
  followingProfile: Profile | null;
}

export const initialState: ProfileState = {
  profile: null,
  followingProfile: null
};

export function profileReducer(
  state = initialState,
  action: ProfileActions
): ProfileState {
  switch (action.type) {
    case ProfileActionTypes.ProfileLoadSuccess:
      return {
        ...state,
        profile: action.payload.profile
      };
    case ProfileActionTypes.ProfileClear:
      return {
        ...state,
        profile: null
      };
    case ProfileActionTypes.ProfileToggleFollowingRequest:
      return {
        ...state,
        profile: { ...state.profile, isFollowLoading: true }
      };
    case ProfileActionTypes.ProfileToggleFollowingFail:
      return {
        ...state,
        profile: { ...state.profile, isFollowLoading: false }
      };
    case ProfileActionTypes.ProfileToggleFollowingSuccess:
      return {
        ...state,
        profile: action.payload.profile
      };
    case ProfileActionTypes.SetFollowingProfile:
      return {
        ...state,
        followingProfile: action.payload.profile
      };
    case ProfileActionTypes.ClearFollowingProfile:
      return {
        ...state,
        followingProfile: null
      };
    default:
      return state;
  }
}
