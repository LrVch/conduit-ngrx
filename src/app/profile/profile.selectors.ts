import { createSelector, createFeatureSelector } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { ProfileState } from './profile.reducer';

export const selectProfileState = createFeatureSelector<AppState, ProfileState>('profile');

export const selectProfile = createSelector(
  selectProfileState,
  (profileState: ProfileState) => profileState.profile
);

export const selectProfileUsername = createSelector(
  selectProfileState,
  (profileState: ProfileState) => profileState.profile && profileState.profile.username
);

export const selectFollowingProfile = createSelector(
  selectProfileState,
  (profileState: ProfileState) => profileState.followingProfile
);
