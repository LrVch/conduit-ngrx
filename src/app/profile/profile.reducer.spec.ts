import * as fromProfile from './profile.reducer';
import {
  ProfileLoadSuccess,
  ProfileClear,
  ProfileToggleFollowingRequest,
  ProfileToggleFollowingFail,
  ProfileToggleFollowingSuccess,
  SetFollowingProfile,
  ClearFollowingProfile
} from './profile.actions';
import { getProfile } from '../lib/testing';

describe('Editor Reducer', () => {
  describe('State changes', () => {
    const profile = getProfile();
    const initialProfileState: fromProfile.ProfileState =
      fromProfile.initialState;

    it('should have an initial state', () => {
      const state = fromProfile.profileReducer(initialProfileState, {
        type: '@@init'
      } as any);

      expect(state).toBe(initialProfileState);
    });

    it('should set pofile on "ProfileLoadSuccess"', () => {
      const init = { type: '@@init' } as any;
      const action = new ProfileLoadSuccess({ profile });

      const state = [init, action].reduce(
        fromProfile.profileReducer,
        initialProfileState
      );

      expect(state).toMatchSnapshot();
    });

    it('should remove pofile on "ProfileClear"', () => {
      const init = { type: '@@init' } as any;
      const action = new ProfileClear();

      const state = [init, action].reduce(
        fromProfile.profileReducer,
        initialProfileState
      );

      expect(state).toMatchSnapshot();
    });

    it('should set isFollowLoading as true in profile on "ProfileToggleFollowingRequest"', () => {
      const init = { type: '@@init' } as any;
      const action = new ProfileToggleFollowingRequest({ profile });

      const state = [init, action].reduce(
        fromProfile.profileReducer,
        initialProfileState
      );

      expect(state).toMatchSnapshot();
    });

    it('should set isFollowLoading as false in profile on "ProfileToggleFollowingFail"', () => {
      const init = { type: '@@init' } as any;
      const action = new ProfileToggleFollowingFail();

      const state = [init, action].reduce(
        fromProfile.profileReducer,
        initialProfileState
      );

      expect(state).toMatchSnapshot();
    });

    it('should set profile on "ProfileToggleFollowingSuccess"', () => {
      const init = { type: '@@init' } as any;
      const action = new ProfileToggleFollowingSuccess({ profile });

      const state = [init, action].reduce(
        fromProfile.profileReducer,
        initialProfileState
      );

      expect(state).toMatchSnapshot();
    });

    it('should set followingProfile on "SetFollowingProfile"', () => {
      const init = { type: '@@init' } as any;
      const action = new SetFollowingProfile({ profile });

      const state = [init, action].reduce(
        fromProfile.profileReducer,
        initialProfileState
      );

      expect(state).toMatchSnapshot();
    });

    it('should remove followingProfile on "ClearFollowingProfile"', () => {
      const init = { type: '@@init' } as any;
      const action = new ClearFollowingProfile();

      const state = [init, action].reduce(
        fromProfile.profileReducer,
        initialProfileState
      );

      expect(state).toMatchSnapshot();
    });
  });
});
