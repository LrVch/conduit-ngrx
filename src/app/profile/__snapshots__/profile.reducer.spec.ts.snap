// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Reducer State changes should remove followingProfile on "ClearFollowingProfile" 1`] = `
Object {
  "followingProfile": null,
  "profile": null,
}
`;

exports[`Editor Reducer State changes should remove pofile on "ProfileClear" 1`] = `
Object {
  "followingProfile": null,
  "profile": null,
}
`;

exports[`Editor Reducer State changes should set followingProfile on "SetFollowingProfile" 1`] = `
Object {
  "followingProfile": Object {
    "bio": "bio",
    "following": false,
    "image": "image",
    "isFollowLoading": false,
    "username": "username",
  },
  "profile": null,
}
`;

exports[`Editor Reducer State changes should set isFollowLoading as false in profile on "ProfileToggleFollowingFail" 1`] = `
Object {
  "followingProfile": null,
  "profile": Object {
    "isFollowLoading": false,
  },
}
`;

exports[`Editor Reducer State changes should set isFollowLoading as true in profile on "ProfileToggleFollowingRequest" 1`] = `
Object {
  "followingProfile": null,
  "profile": Object {
    "isFollowLoading": true,
  },
}
`;

exports[`Editor Reducer State changes should set pofile on "ProfileLoadSuccess" 1`] = `
Object {
  "followingProfile": null,
  "profile": Object {
    "bio": "bio",
    "following": false,
    "image": "image",
    "isFollowLoading": false,
    "username": "username",
  },
}
`;

exports[`Editor Reducer State changes should set profile on "ProfileToggleFollowingSuccess" 1`] = `
Object {
  "followingProfile": null,
  "profile": Object {
    "bio": "bio",
    "following": false,
    "image": "image",
    "isFollowLoading": false,
    "username": "username",
  },
}
`;
